drivers/serial/uart_rpi_pico_pio.c

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) 2022, Yonatan Schachter
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/uart.h>
+
+#include <zephyr/drivers/misc/pio_rpi_pico/pio_rpi_pico.h>
+
+#include <hardware/pio.h>
+#include <hardware/clocks.h>
+
+#define DT_DRV_COMPAT raspberrypi_pico_uart_pio
+
+#define CYCLES_PER_BIT 8
+#define SIDESET_BIT_COUNT 2
+#define TX_STATE_MACHINE_IDX 0
+#define RX_STATE_MACHINE_IDX 1
+
+struct pio_uart_config {
+	const struct device *piodev;
+	const uint32_t tx_pin;
+	const uint32_t rx_pin;
+	uint32_t baudrate;
+	const struct pinctrl_dev_config *pcfg;
+};
+
+RPI_PICO_PIO_DECLARE_PROGRAM(uart_tx, 0, 3,
+		/* .wrap_target */
+	0x9fa0, /*  0: pull   block           side 1 [7]  */
+	0xf727, /*  1: set    x, 7            side 0 [7]  */
+	0x6001, /*  2: out    pins, 1                     */
+	0x0642, /*  3: jmp    x--, 2                 [6]  */
+		/* .wrap */
+);
+
+RPI_PICO_PIO_DECLARE_PROGRAM(uart_rx, 0, 8,
+		/*  .wrap_target */
+	0x2020, /*  0: wait   0 pin, 0                    */
+	0xea27, /*  1: set    x, 7                   [10] */
+	0x4001, /*  2: in     pins, 1                     */
+	0x0642, /*  3: jmp    x--, 2                 [6]  */
+	0x00c8, /*  4: jmp    pin, 8                      */
+	0xc014, /*  5: irq    nowait 4 rel                */
+	0x20a0, /*  6: wait   1 pin, 0                    */
+	0x0000, /*  7: jmp    0                           */
+	0x8020, /*  8: push   block                       */
+		/*  .wrap */
+);
+
+static void pio_uart_tx_init(PIO pio, uint32_t sm, uint32_t pin_tx, float div)
+{
+	uint32_t offset;
+	pio_sm_config sm_config;
+
+	offset = pio_add_program(pio, &RPI_PICO_PIO_GET_PROGRAM(uart_tx));
+	sm_config = pio_get_default_sm_config();
+
+	sm_config_set_sideset(&sm_config, SIDESET_BIT_COUNT, true, false);
+	sm_config_set_out_shift(&sm_config, true, false, 0);
+	sm_config_set_out_pins(&sm_config, pin_tx, 1);
+	sm_config_set_sideset_pins(&sm_config, pin_tx);
+	sm_config_set_fifo_join(&sm_config, PIO_FIFO_JOIN_TX);
+	sm_config_set_clkdiv(&sm_config, div);
+	sm_config_set_wrap(&sm_config,
+			   offset + RPI_PICO_PIO_GET_WRAP_TARGET(uart_tx),
+			   offset + RPI_PICO_PIO_GET_WRAP(uart_tx));
+
+	pio_sm_set_pins_with_mask(pio, sm, BIT(pin_tx), BIT(pin_tx));
+	pio_sm_set_pindirs_with_mask(pio, sm, BIT(pin_tx), BIT(pin_tx));
```

Comment:
if this is the only thing required for pins, why can't pinctrl driver do this if function is PIO0/1?

Reply:
There are more than just those two functions that use the pin - e.g. `sm_config_set_out_pins`. It is possible to define the configuration of each state machine in dts and have the generic driver apply that configuration. However, I think this would greatly complicate the dts.